### ISSUE TITLE
[core][nightly] many_nodes_actor_test_on_v2.aws failed 

### ISSUE THREAD

**Issue by rickyyx** [ray-project/ray#34635]
### What happened + What you expected to happen

The test failed with timeout - seems to be an infra issue to me: 
- No logs (from job submit & through ray logs download) are available 

### Versions / Dependencies

master

### Reproduction script

https://buildkite.com/ray-project/release-tests-branch/builds/1568#01879142-b94c-48b2-9af8-bac773cd78b0

### Issue Severity

None

**Comment by can-anyscale** [comment]
I ran two bisect https://buildkite.com/ray-project/release-tests-bisect/builds/98#_ and https://buildkite.com/ray-project/release-tests-bisect/builds/101#_, both blame https://github.com/ray-project/ray/commit/7c9da5c2935de95976366064bac598c52eeb30dc (committed 2 weeks ago)

**Comment by rickyyx** [comment]
Thanks @can-anyscale  That's possible - will look into this bisecting and the potential root caused commit later next week. 


What's weird to me is this failure itself doesn't seem to have any logs generated?

**Comment by rickyyx** [comment]
https://github.com/anyscale/product/issues/19989

**Comment by iycheng** [comment]
The issue might be due to pubsub. The theory here is that, 

- Raylet report failure
- CoreWorker close slower than before
- GCS's closed the long polling.
- CoreWorker sent another pubsub request
- GCS got the long polling again.

Thus leak in the end.

Going to verify this theory.

**Comment by iycheng** [comment]
```
581e298771cdbeb7e50c61e3e7bdf14028d07259cf3ab93db3aa91b7 581e298771cdbeb7e50c61e3e7bdf14028d07259cf3ab93db3aa91b7 Adding
581e298771cdbeb7e50c61e3e7bdf14028d07259cf3ab93db3aa91b7 faf9864ff731dc86d96007303211de3a7df3a66e8b979cc47ba055bd Adding
581e298771cdbeb7e50c61e3e7bdf14028d07259cf3ab93db3aa91b7 581e298771cdbeb7e50c61e3e7bdf14028d07259cf3ab93db3aa91b7 Remove
581e298771cdbeb7e50c61e3e7bdf14028d07259cf3ab93db3aa91b7 faf9864ff731dc86d96007303211de3a7df3a66e8b979cc47ba055bd Remove
581e298771cdbeb7e50c61e3e7bdf14028d07259cf3ab93db3aa91b7 faf9864ff731dc86d96007303211de3a7df3a66e8b979cc47ba055bd Adding
581e298771cdbeb7e50c61e3e7bdf14028d07259cf3ab93db3aa91b7 6179df5e269ae1961183d8f6de985f0c33bd64d613e4a071a9b92b85 Adding
581e298771cdbeb7e50c61e3e7bdf14028d07259cf3ab93db3aa91b7 b82ce2de937dc9c429842407abcaa8ada3d0aacb778d22334bd80cbb Adding
```
I think the theory is correct here.

Basically, some subscriber got removed, but before the worker stopped, it sends another subscription to the node.

**Comment by iycheng** [comment]
The leak PR is merged. Start a new run https://console.anyscale-staging.com/o/anyscale-internal/jobs/prodjob_ztbbc72u62tfzvrkypgbdqpq7c

**Comment by can-anyscale** [comment]
Look like it's still failing :(

**Comment by iycheng** [comment]
```
2023-05-05 13:13:13,234 ERROR reporter_agent.py:1112 -- Error publishing node physical stats.
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/modules/reporter/reporter_agent.py", line 1092, in _perform_iteration
    timeout=GCS_RPC_TIMEOUT_SECONDS,
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/gcs_utils.py", line 167, in wrapper
    return await f(self, *args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/gcs_utils.py", line 280, in internal_kv_get
    reply = await self._kv_stub.InternalKVGet(req, timeout=timeout)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/grpc/aio/_call.py", line 291, in __await__
    self._cython_call._status)
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
        status = StatusCode.DEADLINE_EXCEEDED
        details = "Deadline Exceeded"
        debug_error_string = "UNKNOWN:Deadline Exceeded {created_time:"2023-05-05T13:13:13.234129778-07:00", grpc_status:4}"
>
2023-05-05 13:13:18,736 ERROR reporter_agent.py:1112 -- Error publishing node physical stats.
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/modules/reporter/reporter_agent.py", line 1092, in _perform_iteration
    timeout=GCS_RPC_TIMEOUT_SECONDS,
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/gcs_utils.py", line 167, in wrapper
    return await f(self, *args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/gcs_utils.py", line 280, in internal_kv_get
    reply = await self._kv_stub.InternalKVGet(req, timeout=timeout)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/grpc/aio/_call.py", line 291, in __await__
    self._cython_call._status)
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
        status = StatusCode.DEADLINE_EXCEEDED
        details = "Deadline Exceeded"
        debug_error_string = "UNKNOWN:Deadline Exceeded {created_time:"2023-05-05T13:13:18.736149177-07:00", grpc_status:4}"
>
```

The dashboard agent failed somehow because it failed to talk with GCS. I think don't make agent fate sharing with raylet is critical. Ant seems don't have the bandwidth anymore, maybe we should just do this. cc @rkooo567 

Still checking what caused the regression. GCS seem ok. and the raylet failure is because of agent failure. Agent failure is because of failing to talk to GCS.

**Comment by iycheng** [comment]
The PR got reverted and the revert-revert is here https://github.com/ray-project/ray/pull/35091

I'll test once the wheel is built.

**Comment by rkooo567** [comment]
Seems like this test hasn't run for a while. We should follow up with @can-anyscale to verify why it hasn't run

**Comment by SongGuyang** [comment]
> The dashboard agent failed somehow because it failed to talk with GCS. I think don't make agent fate sharing with raylet is critical. @SongGuyang does your team still have bandwidth for this? I think if no bandwidth, maybe we should cover this cc @rkooo567

Sorry for the late update. We already restarted this work. The pr will be created in a few days.

**Comment by can-anyscale** [comment]
@rkooo567: this test in on a nightly 3x schedule; last time it ran I think it's still failing

**Comment by rkooo567** [comment]
Is it consistently failing? Maybe we should just reduce the num_nodes to something like 1000 instead of making it keep failing

**Comment by can-anyscale** [comment]
@rkooo567: it has been failing for 3 weeks, we should prioritize this to avoid delay the upcoming release

**Comment by rkooo567** [comment]
@iycheng if https://github.com/ray-project/ray/pull/35091 doesn't solve the issue (can you verify it by running the release test from the PR?), I think the best way is to reduce our scalability envelope and revisit.

**Comment by can-anyscale** [comment]
> I think the best way is to reduce our scalability envelope and revisit.

Can we give this a try @iycheng , @rkooo567 , thankks

**Comment by iycheng** [comment]
> @iycheng if #35091 doesn't solve the issue (can you verify it by running the release test from the PR?), I think the best way is to reduce our scalability envelope and revisit.

@rkooo567  I think it's a regression then. Shouldn't we fix the regression? The test has been running healthy for more than a month. Then later due to the file descriptor leak, it's broken.

What's worse, during the regression, there is another bug which prevent us from bisecting. This means even fixing the FD leak, it's not enough.

The goal should be to identify the why it's broken and fix it in 2.5. (2.3 we are able to run this test, 2.4 it regressed) not just that we are able to run the test.

**Comment by iycheng** [comment]
After applying the fixing of FD issues, the new logs shows:
```
[2023-05-13 15:37:52,450 I 247 247] (gcs_server) gcs_actor_scheduler.cc:521: Retry creating actor e9bc491a509697c0d7d1f4a803000000 on worker a4bd10b252cd36fcf7a7c62d55e2d1fa105eaff53c76e5c80b458052 at node 566a2ba4
df0619b34c3a262af92f980b70bad018fd0d99799827b952, job id = 03000000
[2023-05-13 15:37:52,450 I 247 247] (gcs_server) gcs_actor_scheduler.cc:445: Start creating actor e9bc491a509697c0d7d1f4a803000000 on worker a4bd10b252cd36fcf7a7c62d55e2d1fa105eaff53c76e5c80b458052 at node 566a2ba4
df0619b34c3a262af92f980b70bad018fd0d99799827b952, job id = 03000000
[2023-05-13 15:37:52,666 I 247 247] (gcs_server) gcs_actor_scheduler.cc:521: Retry creating actor e9bc491a509697c0d7d1f4a803000000 on worker a4bd10b252cd36fcf7a7c62d55e2d1fa105eaff53c76e5c80b458052 at node 566a2ba4
df0619b34c3a262af92f980b70bad018fd0d99799827b952, job id = 03000000
[2023-05-13 15:37:52,666 I 247 247] (gcs_server) gcs_actor_scheduler.cc:445: Start creating actor e9bc491a509697c0d7d1f4a803000000 on worker a4bd10b252cd36fcf7a7c62d55e2d1fa105eaff53c76e5c80b458052 at node 566a2ba4
df0619b34c3a262af92f980b70bad018fd0d99799827b952, job id = 03000000
[2023-05-13 15:37:52,940 I 247 247] (gcs_server) gcs_actor_scheduler.cc:521: Retry creating actor e9bc491a509697c0d7d1f4a803000000 on worker a4bd10b252cd36fcf7a7c62d55e2d1fa105eaff53c76e5c80b458052 at node 566a2ba4
df0619b34c3a262af92f980b70bad018fd0d99799827b952, job id = 03000000
[2023-05-13 15:37:52,940 I 247 247] (gcs_server) gcs_actor_scheduler.cc:445: Start creating actor e9bc491a509697c0d7d1f4a803000000 on worker a4bd10b252cd36fcf7a7c62d55e2d1fa105eaff53c76e5c80b458052 at node 566a2ba4
df0619b34c3a262af92f980b70bad018fd0d99799827b952, job id = 03000000
[2023-05-13 15:37:53,195 I 247 247] (gcs_server) gcs_actor_scheduler.cc:521: Retry creating actor e9bc491a509697c0d7d1f4a803000000 on worker a4bd10b252cd36fcf7a7c62d55e2d1fa105eaff53c76e5c80b458052 at node 566a2ba4
df0619b34c3a262af92f980b70bad018fd0d99799827b952, job id = 03000000
```

Seems worker failure is not handled in GCS.

The crash is in destruction:

```
[2023-05-13 15:35:12,273 I 671 671] core_worker_process.cc:148: Destructing CoreWorkerProcessImpl. pid: 671
[2023-05-13 15:35:12,273 I 671 671] io_service_pool.cc:47: IOServicePool is stopped.
[2023-05-13 15:35:12,310 I 671 671] core_worker.cc:615: Core worker is destructed
[2023-05-13 15:35:12,360 E 671 671] logging.cc:104: Stack trace:
 /home/ray/anaconda3/lib/python3.7/site-packages/ray/_raylet.so(+0xdbdcaa) [0x7f3a1fa0acaa] ray::operator<<()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/_raylet.so(+0xdc0468) [0x7f3a1fa0d468] ray::TerminateHandler()
/home/ray/anaconda3/bin/../lib/libstdc++.so.6(+0xb135a) [0x7f3a1eae035a] __cxxabiv1::__terminate()
/home/ray/anaconda3/bin/../lib/libstdc++.so.6(+0xb13c5) [0x7f3a1eae03c5]
/home/ray/anaconda3/lib/python3.7/site-packages/ray/_raylet.so(+0x7059a1) [0x7f3a1f3529a1]
/home/ray/anaconda3/lib/python3.7/site-packages/ray/_raylet.so(_ZN3ray4core6worker19TaskEventBufferImplD0Ev+0x12) [0x7f3a1f3529c2] ray::core::worker::TaskEventBufferImpl::~TaskEventBufferImpl()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/_raylet.so(_ZN3ray4core10CoreWorkerD1Ev+0x50) [0x7f3a1f2ccf90] ray::core::CoreWorker::~CoreWorker()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/_raylet.so(+0x5675aa) [0x7f3a1f1b45aa] std::_Sp_counted_base<>::_M_release()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/_raylet.so(_ZN3ray4core21CoreWorkerProcessImplD1Ev+0x101) [0x7f3a1f3094d1] ray::core::CoreWorkerProcessImpl::~CoreWorkerProcessImpl()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/_raylet.so(_ZN3ray4core17CoreWorkerProcess12HandleAtExitEv+0x29) [0x7f3a1f3096a9] ray::core::CoreWorkerProcess::HandleAtExit()
/lib/x86_64-linux-gnu/libc.so.6(+0x468a7) [0x7f3a2061e8a7]
/lib/x86_64-linux-gnu/libc.so.6(on_exit+0) [0x7f3a2061ea60] on_exit
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xfa) [0x7f3a205fc08a] __libc_start_main
ray::IDLE() [0x543fce]
```

**Comment by iycheng** [comment]
This PR should fix this test https://github.com/ray-project/ray/pull/35320

It's pretty trick here and the root cause is still because we didn't report the work failure correctly.

I'll wait for the CI tests.

https://console.anyscale-staging.com/o/anyscale-internal/jobs/prodjob_7ik9xza3l1jelj9qih8tghkp66

**Comment by rkooo567** [comment]
cc @rickyyx the crash seems relevant to the task backend? @iycheng does the same error happen if you set RAY_task_events_report_interval_ms=0?

**Comment by rkooo567** [comment]
Hmm it seems like TaskEventBuffer doesn't really have any destructor...

**Comment by iycheng** [comment]
@rkooo567 I think the crash showing in the log is orthogonal to the test failure but still something good to fix. This test failed because the FD leak. The liveness check is not great in my first fix.

**Comment by rickyyx** [comment]
> cc @rickyyx the crash seems relevant to the task backend? @iycheng does the same error happen if you set RAY_task_events_report_interval_ms=0?

Likely due to the non-graceful exit (no Stop() being called) when this happens? But yeah, I guess this could be fixed.

**Comment by can-anyscale** [comment]
@iycheng: this test is passing on master now, can you confirm and close the issue if that's true. Thanks

**Comment by can-anyscale** [comment]
@iycheng : this test is now failing again with a different reason https://buildkite.com/ray-project/release-tests-branch/builds/1657#01882a60-6091-4c35-99ba-52dc33df0c93

**Comment by iycheng** [comment]
It also break one CI test. It got reverted. Taking another look.

**Comment by iycheng** [comment]
The failure is because sometimes, worker failure is failed to be reported.

**Comment by iycheng** [comment]
When the test failed, there are other things happened which increase the fd usage of a core worker. That's why previous my test passed and later it's merged, it failed.

For short term, we'll update the test: https://github.com/ray-project/ray/pull/35546

Successful run here:
https://console.anyscale-staging.com/o/anyscale-internal/jobs/prodjob_7c11at2fr37mqi4xrnybedquxp

```
Result: {
  "many_nodes_actor_tests_10000": {
    "actor_launch_time": 2.7677583710000135,
    "actor_ready_time": 64.35485191700002,
    "total_time": 67.12261028800003,
    "num_actors": 10000,
    "success": "1",
    "throughput": 148.9810952984909
  },
  "many_nodes_actor_tests_15000": {
    "actor_launch_time": 3.516498482999964,
    "actor_ready_time": 195.170059293,
    "total_time": 198.68655777599997,
    "num_actors": 15000,
    "success": "1",
    "throughput": 75.49579683649792
  }
}
```

I'll rerun the tests in the PR before merge.

After that, I'll check which pr increase the fds and add tests to prevent that from happening again.

**Comment by can-anyscale** [comment]
BTW, @iycheng, 2.5 release branch still has your previous fix that was reverted in master, do we need to revert it in release branch as well? Thanks

**Comment by can-anyscale** [comment]
Since this is a release-blocker issue, please **close it only after the cherry pick fix is merged into 2.5 release branch**. 

Please add @ArturNiederfahrenhorst as one of the reviewer of the fix as well for tracking purpose. Thankks!

**Comment by iycheng** [comment]
@can-anyscale yes, we should cherry-pick that. Let me create one

**Comment by can-anyscale** [comment]
@iycheng got you, that's great, thank you

**Comment by iycheng** [comment]
After a day's of checking, it turns out because of getting rid of grpcio work, it actually increase the sockets in GCS.

So there are two ways for this test:
- update the scalability envelop and accept the regression
- fix this in some way and keep the metrics.

I have two PRs for both. Both passed the nightly tests.

The fixing one has ci failures. I'll try to fix them.

**Comment by iycheng** [comment]
@can-anyscale the fix has been merged. Feel free to verify it in your end once the master wheel is built.

The successful run: https://console.anyscale-staging.com/o/anyscale-internal/jobs/prodjob_ywxup58wj76i8567e52l3uiijb

**Comment by can-anyscale** [comment]
@iycheng w00h00 thanks

**Comment by iycheng** [comment]
Triggered another run through the buildkite on master branch and it passed: https://buildkite.com/ray-project/release-tests-branch/builds/1692#0188500b-3c34-4fc8-8cbe-2566859c716c
